File: tests/fixtures/cpp_repo/src/services/user_service.cpp
#include "api/services/user_service.hpp"

#include "api/error/api_error.hpp"
#include "api/repository/user_repository.hpp"
#include "api/services/auth_service.hpp"

#include <algorithm>
#include <chrono>
#include <utility>

namespace api {

namespace {

std::int64_t now_seconds() {
    using namespace std::chrono;
    return duration_cast<seconds>(system_clock::now().time_since_epoch()).count();
}

}  // namespace

UserService::UserService(std::shared_ptr<UserRepository> users,
                         std::shared_ptr<AuthService> auth)
    : users_(std::move(users)), auth_(std::move(auth)) {}

UserResponse UserService::create_user(const CreateUserRequest& request) {
    if (users_->exists_by_email(request.email)) {
        throw ApiError::conflict("email already taken");
    }
    User user{};
    user.email = request.email;
    user.username = request.username;
    user.password_hash = auth_->encode_password(request.password);
    user.created_at = now_seconds();
    User saved = users_->save(user);
    return UserResponse::from_entity(saved);
}

UserResponse UserService::get_user_by_id(std::int64_t id) {
    auto user = users_->find_by_id(id);
    if (!user.has_value()) {
        throw ApiError::not_found("user not found");
    }
    return UserResponse::from_entity(*user);
}

std::vector<UserResponse> UserService::list_users(int page, int page_size) {
    int safe_page = std::max(0, page);
    int safe_size = std::clamp(page_size, 1, kMaxPageSize);
    auto rows = users_->find_all_paged(safe_page, safe_size);
    std::vector<UserResponse> out;
    out.reserve(rows.size());
    for (const auto& user : rows) {
        out.push_back(UserResponse::from_entity(user));
    }
    return out;
}

UserResponse UserService::update_user(std::int64_t id,
                                      const UpdateUserRequest& patch) {
    auto user = users_->find_by_id(id);
    if (!user.has_value()) {
        throw ApiError::not_found("user not found");
    }
    if (patch.email.has_value()) {
        user->email = *patch.email;
    }
    if (patch.username.has_value()) {
        user->username = *patch.username;
    }
    if (patch.password.has_value()) {
        user->password_hash = auth_->encode_password(*patch.password);
    }
    User saved = users_->update(*user);
    return UserResponse::from_entity(saved);
}

void UserService::delete_user(std::int64_t id) {
    if (!users_->exists_by_id(id)) {
        throw ApiError::not_found("user not found");
    }
    users_->delete_by_id(id);
}

}  // namespace api

File: tests/fixtures/java_repo/src/main/java/com/example/api/services/UserService.java
package com.example.api.services;

import com.example.api.dto.CreateUserRequest;
import com.example.api.dto.UpdateUserRequest;
import com.example.api.dto.UserResponse;
import com.example.api.exceptions.ApiException;
import com.example.api.models.User;
import com.example.api.repositories.UserRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

/**
 * UserService — CRUD operations for {@link User} aggregates.
 *
 * <p>Hashes passwords via {@link AuthService#encodePassword}, enforces
 * email uniqueness, and clamps pagination parameters so the controller
 * layer can hand raw query strings to us safely.
 */
@Service
public class UserService {

    private static final int MAX_PAGE_SIZE = 100;

    private final UserRepository userRepository;
    private final AuthService authService;

    public UserService(UserRepository userRepository, AuthService authService) {
        this.userRepository = userRepository;
        this.authService = authService;
    }

    /**
     * createUser hashes the request's plaintext password, persists a new
     * {@link User}, and returns the public DTO.
     */
    @Transactional
    public UserResponse createUser(CreateUserRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new ApiException(HttpStatus.CONFLICT, "email already taken");
        }
        String hashed = authService.encodePassword(request.getPassword());
        User user = new User(request.getEmail(), request.getUsername(), hashed);
        User saved = userRepository.save(user);
        return UserResponse.fromEntity(saved);
    }

    /**
     * getUserById returns the user with the given id or 404s.
     */
    public UserResponse getUserById(Long id) {
        User user = userRepository.findById(id)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "user not found"));
        return UserResponse.fromEntity(user);
    }

    /**
     * listUsers returns a page of users wrapped in DTOs.
     */
    public List<UserResponse> listUsers(int page, int pageSize) {
        int safePage = Math.max(0, page);
        int safeSize = Math.min(MAX_PAGE_SIZE, Math.max(1, pageSize));
        Pageable pageable = PageRequest.of(safePage, safeSize);
        Page<User> rows = userRepository.findAllPaged(pageable);
        return rows.stream().map(UserResponse::fromEntity).toList();
    }

    /**
     * updateUser applies the non-null fields of {@code patch} to the
     * tracked entity and re-hashes the password if supplied.
     */
    @Transactional
    public UserResponse updateUser(Long id, UpdateUserRequest patch) {
        User user = userRepository.findById(id)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "user not found"));
        if (patch.getEmail() != null) {
            user.setEmail(patch.getEmail());
        }
        if (patch.getUsername() != null) {
            user.setUsername(patch.getUsername());
        }
        if (patch.getPassword() != null) {
            user.setPasswordHash(authService.encodePassword(patch.getPassword()));
        }
        return UserResponse.fromEntity(user);
    }

    /**
     * deleteUser removes the user with the given id or 404s.
     */
    @Transactional
    public void deleteUser(Long id) {
        if (!userRepository.existsById(id)) {
            throw new ApiException(HttpStatus.NOT_FOUND, "user not found");
        }
        userRepository.deleteById(id);
    }
}
